src/engraving/rendering/dev/chordlayout.cpp

```diff
@@ -2640,13 +2572,113 @@ void ChordLayout::layoutChords3(const MStyle& style, const std::vector<Chord*>&
+        for (Note* note : chordNotes) {
+            double noteX = 0.0;
+            if (note->ldata()->mirror()) {
+                if (_up) {
+                    noteX = chord->stemPosX() - overlapMirror;
+                } else {
+                    noteX = -note->headBodyWidth() + overlapMirror;
+                }
+            } else if (_up) {
+                noteX = chord->stemPosX() - note->headBodyWidth();
+            }
+
+            double ny = (note->line() + stepOffset) * stepDistance;
+            if (note->ldata()->pos().y() != ny) {
+                note->mutldata()->setPosY(ny);
+                if (stem) {
+                    TLayout::layoutStem(chord->stem(), chord->stem()->mutldata(), ctx.conf());
+                    if (chord->hook()) {
+                        chord->hook()->mutldata()->setPosY(chord->stem()->flagPosition().y());
+                    }
+                }
+            }
+            note->mutldata()->setPosX(noteX);
+
+            // find leftmost non-mirrored note to set as X origin for accidental layout
+            // a mirrored note that extends to left of segment X origin
+            // will displace accidentals only if there is conflict
+            double segmentX = noteX + chord->x();     // segment-relative X position of note
+            if (note->ldata()->mirror() && !_up && segmentX < -note->headBodyWidth() / 2) {
+                leftNotes.push_back(note);
+            } else if (segmentX < lx) {
+                lx = segmentX;
+            }
+
+            minDotPosX = std::max(minDotPosX, noteX + note->headBodyWidth() + chord->pos().x());
+        }
+
+        //---------------------------------------------------
+        //    layout dots simply
+        //     we will check for conflicts after all the notes have been processed
+        //---------------------------------------------------
+
+        if (chord->dots()) {
+            bool hasNoVisibleDots = true;
+            bool chordAlreadyHasDots = false;
+            for (Note* n : chord->notes()) {
+                if (!hasNoVisibleDots) {
+                    break;
+                }
+                for (NoteDot* dot : n->dots()) {
+                    chordAlreadyHasDots = true;
+                    if (dot->visible()) {
+                        hasNoVisibleDots = false;
+                        break;
+                    }
+                }
+            }
+            if (hasNoVisibleDots && chordAlreadyHasDots) {
+                chordsWithNoVisibleDots.push_back(chord);
+                chord->setDotPosX(minDotPosX);
+            } else if (chord->up()) {
+                upDotPosX = std::max(upDotPosX, minDotPosX);
+            } else {
+                downDotPosX = std::max(downDotPosX, minDotPosX);
+            }
+        }
+    }
+
     if (conflict) {
         double maxPosX = std::max(upDotPosX, downDotPosX);
         for (Chord* chord : chords) {
+            if (mu::contains(chordsWithNoVisibleDots, chord)) {
```

Comment:
And here, instead of creating a vector to hold these chords, you can just call `chord->hasDotsAllInvisible()` again (performance won't be affected I think)